enhancements/installer/credentials-management-outside-openshift-cluster.md

```diff
@@ -0,0 +1,130 @@
+---
+title: Credentials Management outside OpenShift Cluster
+authors:
+  - "@akhil-rane"
+reviewers:
+  - "@abhinavdahiya"
+  - "@dgoodwin"
+  - "@joelddiaz"
+approvers:
+  - "@derekwaynecarr"
+  - "@sdodson"
+creation-date: 2021-03-25
+last-updated: 2021-03-25
+status: provisional
+---
+
+# Credentials Management outside OpenShift Cluster
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+The intent of this enhancement is to take the process of credentials management outside the OpenShift cluster for new 
+platforms. We propose to make *manual* mode as default for clusters on new platforms. 
+
+## Motivation
+
+The main motivation behind this enhancement is to satisfy the customer requirement to follow the best security practices 
+by not storing admin credentials inside the cluster
+
+### Goals
+
+As part of this enhancement we plan to do the following:
+* Set *manual* as default/preferred credentials mode for OpenShift
+* Not store admin level cloud credentials inside a cluster
+* Provide a tooling design for credentials management outside the cluster
+
+### Non-Goals
+
+* We do not plan to go into design details on how this tool will be ported to cloud providers other than AWS.
+
+## Proposal
+
+Currently, *mint* is a default credentials mode for OpenShift. In this mode we run OpenShift installer with an admin 
+level cloud credential. The admin credential is stored in kube-system namespace and then used by the cloud credential 
+operator to process the CredentialRequests in the cluster and create new users with fine-grained permissions.
+The customers have reported that the need to store admin credentials inside the cluster is a major disadvantage. Based 
+on the feedback, we propose to run credentials related setup outside a cluster and then start installation process in a 
+*manual* mode without providing admin credentials to the installer.
+
+### User Stories
+
+#### Story 1
+As a OpenShift cluster administrator I should be able to extract the CredentialsRequest manifests from the release image
+and create credentials in the underlying cloud provider. 
+
+#### Story 2
+As a OpenShift cluster administrator I should be able to extract the CredentialsRequest manifests from the release image 
+and create Kubernetes Secrets in to satisfy all in-cluster component's CredentialsRequests.
+
+#### Story 3
+As a OpenShift cluster administrator I should be able to install a OpenShift cluster, without providing admin credentials
+to the installer, by injecting precreated Kubernetes Secrets.
```

Comment:
How do we envision a user doing this injection? Is it going to be through custom manifests added manually by the user to {install-dir}/manifests? I am concerned about the complexity of our recommended flow. Is it practical for the installer to do the job of the `ccoctl` tool?

Reply:
Yes manual mode today involved injecting via custom manifests, but ccoctl will generate everything for you and drop the files if desired.

Integrating right into openshift-install is interesting. If that is on the table that could be very useful, but we'd still want a ccoctl (or oc adm?) entrypoint for on-going maintenance I think.

Reply:
Yes, for on-going maintenance. But it seems like a much better user experience to have the installer do the ccoctl work during install, rather than the preferred installation path requiring multiple steps with multiple tools.

Reply:
Agree with above for credentials that are scoped to a single cluster.

Reply:
@staebler do we want to outline how that could work in this enhancement or leave it for later? If so, how could it work? :) Are you thinking new openshift-install subcommand, something integrated into installconfig, or other?

Reply:
I would think that it would be the default behavior of the installer. There would not be a sub-command. If the user wanted to opt out of having the installer creates the roles and secret manifests, then there would be an option in the install config to tell the installer to not create them. Could that be a new value for `credentialsMode`?

Reply:
IIUC same thing is happening when we set `credentialsMode` to manual. We might need to add a different (default) mode where in installer does ccoctl work as part of install.